xarray/coding/variables.py

```diff
@@ -298,18 +298,61 @@ def _check_fill_values(attrs, name, dtype):
     return raw_fill_dict, encoded_fill_values
 
 
+def _convert_unsigned_fill_value(
+    name: T_Name,
+    data: Any,
+    unsigned: str,
+    raw_fill_value: Any,
+    encoded_fill_values: set,
+) -> Any:
+    if data.dtype.kind == "i":
+        if unsigned == "true":
+            unsigned_dtype = np.dtype(f"u{data.dtype.itemsize}")
+            transform = partial(np.asarray, dtype=unsigned_dtype)
+            if raw_fill_value is not None:
+                new_fill = np.array(raw_fill_value, dtype=data.dtype)
+                encoded_fill_values.remove(raw_fill_value)
+                # use view here to prevent OverflowError
+                encoded_fill_values.add(new_fill.view(unsigned_dtype).item())
+            data = lazy_elemwise_func(data, transform, unsigned_dtype)
+    elif data.dtype.kind == "u":
+        if unsigned == "false":
+            signed_dtype = np.dtype(f"i{data.dtype.itemsize}")
+            transform = partial(np.asarray, dtype=signed_dtype)
+            data = lazy_elemwise_func(data, transform, signed_dtype)
+            if raw_fill_value is not None:
+                new_fill = signed_dtype.type(raw_fill_value)
```

Comment:
This block didn't use @kmuehlbauer's trick for handling overflow by using `.view`. I'm wondering if I need to use that here, but no tests hit it. I've never actually seen `_Unsigned == "false"` in the wild.

Reply:
If I change that here then I think I might be able to shrink this function and do things as "old dtype" and "new dtype" rather than signed versus unsigned.

Reply:
> This block didn't use @kmuehlbauer's trick for handling overflow by using .view. I'm wondering if I need to use that here, but no tests hit it. I've never actually seen _Unsigned == \"false\" in the wild.

That was requested at some point in time for a specific use case. I'll try to dig it up.

Reply:
OK looking at the tests, it looks like this case is not tested (but I'm triple checking). This set of `if`s says that the data on-disk is unsigned, but `_Unsigned` is false which means the user wants signed data in-memory. The tests for `_Unsigned="false"` have signed data on-disk and in-memory so no casting/conversion happens.

Reply:
xref https://github.com/pydata/xarray/pull/4966

Reply:
Well I thought I was being smart and added tests for the `_Unsigned: "false"` case and now things are just all weird. It made me realize I wasn't handling this case in the encoding step, but it also kind of seems like it never was handled or that I have the wrong impression of how that configuration is supposed to be handled. The tests added in that PR @kmuehlbauer don't present the `_FillValue` so I've tried adding that to the backend tests but now non-NC4 backends are complaining about converting uint8 to int8. This coercion was added to solve https://github.com/pydata/xarray/issues/4014 it seems. 

My assumption is that if `_Unsigned: "false"` then the data is saved as uint8 and `_FillValue` should be uint8. But again, I'm not sure why my new tests are even trying to get to int8. I'll do some more testing later tonight hopefully. I'm not sure if it is better to spend a ton of time getting this small functionality working "as expected" or leave it undefined/untested as it is right now.

Reply:
So my main concern and all of your handling in the decode pipeline for casting fill values turns out to not be an issue anymore as the raw fill values turn out to be numpy scalars (`np.uint8`) when they get loaded from the file. Or at least they are for the NetCDF4 cases. So numpy is perfectly happy casting uint8 to int8 and back if they are already numpy scalars. If my new tests cases make sense then I *think* this is fine.